code/modules/minimap/minimap_ui.dm

```diff
@@ -0,0 +1,205 @@
+	var/datum/minimap/minimap_datum
+	///An associative list of data for each minimap marker, so that the UI may read it.
+	var/list/minimap_markers_list
+	///Whether the next call of `"toggle_visibility_all"` will turn all markers opaque or transparent. Does not reflect whether the markers are opaque or transparent.
+	var/markers_visible = TRUE
+
+	New(parent, control_id = null, obj/minimap_controller/minimap_controller, tgui_title, tgui_theme)
+		..(parent, control_id, minimap_controller, tgui_title, tgui_theme)
+		src.minimap_controller = minimap_controller
+		src.minimap_datum = minimap_controller.controlled_minimap.map
+
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if(!ui)
+			ui = new(user, src, "MinimapController")
+			ui.open()
+
+	ui_static_data(mob/user)
+		. = ..()
+		var/list/placable_marker_states = list()
+		var/list/placable_marker_images = list()
+		for(var/icon_state in icon_states('icons/obj/minimap/minimap_markers.dmi'))
+			placable_marker_states.Add(icon_state)
+			var/icon/marker_icon = icon('icons/obj/minimap/minimap_markers.dmi', icon_state)
+			placable_marker_images[icon_state] = icon2base64(marker_icon)
+
+		. += list(
+			"placable_marker_states" = placable_marker_states,
+			"placable_marker_images" = placable_marker_images,
+			"x" = 1,
+			"y" = 1,
+			"icon" = "pin",
+			"image" = placable_marker_images["pin"]
+		)
+
+	ui_data(mob/user)
+		..()
+		minimap_markers_list = list()
+		for (var/atom/target in src.minimap_datum.minimap_markers)
+			var/datum/minimap_marker/marker = src.minimap_datum.minimap_markers[target]
+			if (!marker.on_minimap_z_level)
+				continue
+
+			minimap_markers_list.Add(list(list(
+				"name" = marker.name,
+				"pos" = "[target.x], [target.y]",
+				"visible" = marker.visible,
+				"can_be_deleted" = marker.can_be_deleted_by_player,
+				"marker" = marker,
+				"index" = length(minimap_markers_list) + 1
+			)))
+
+		. = list(
+			"markers_visible" = markers_visible,
+			"selecting_coordinates" = src.minimap_controller.selecting_coordinates,
+			"minimap_markers" = minimap_markers_list
+		)
+
+		if (src.minimap_controller.selected_x && src.minimap_controller.selected_y)
+			. += list(
+				"x" = src.minimap_controller.selected_x,
+				"y" = src.minimap_controller.selected_y
+			)
+			src.minimap_controller.selected_x = null
+			src.minimap_controller.selected_y = null
+
+	ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+		var/mob/user = ui.user
+
+		if(is_incapacitated(user) || !(src.loc in user.equipped_list()))
+			return
+
+		switch (action)
+			if ("reset_scale")
+				src.minimap_controller.reset_scale()
+
+			if ("toggle_visibility_all")
+				if (markers_visible == TRUE)
+					markers_visible = FALSE
+				else
+					markers_visible = TRUE
+
+				src.minimap_controller.toggle_visibility_all(markers_visible)
+
+			if ("toggle_visibility")
+				var/list/list_entry = src.minimap_markers_list[params["index"]]
```

Comment:
Using indices for these as a param (and in the UI) might lead to some race condition things if two people are messing with it at once (consider using a UUID, or even a simply incrementing ID number). Also would want to throw some null checks in when trying to act on a thing from this list, as it may not be on there (as someone else could have deleted it).